File: examples/go/resolve_goto.go
// Resolve a Google /goto URL by reading the Location header.
//
// Google Search results often wrap destinations as:
//
//	https://www.google.com/goto?url=CAES...
//
// The CAES blob is opaque — you cannot decode it. Google only reveals the
// real URL in the HTTP Location header.
//
// Do not follow the redirect (CheckRedirect returns ErrUseLastResponse).
// Try HEAD first, then GET. Google sometimes returns 402 (or another
// non-3xx) with Location: still read the header.
//
// Usage:
//
//	go run ./examples/go 'https://www.google.com/goto?url=CAES...'
package main

import (
	"errors"
	"fmt"
	"net/http"
	"net/url"
	"os"
	"strings"
)

func isGoogleGoto(raw string) bool {
	u, err := url.Parse(raw)
	if err != nil {
		return false
	}
	host := strings.ToLower(u.Hostname())
	return u.Scheme == "https" &&
		(host == "google.com" || host == "www.google.com") &&
		(u.Path == "/goto" || strings.HasPrefix(u.Path, "/goto/"))
}

func resolveGoto(raw string) (string, error) {
	if !isGoogleGoto(raw) {
		return "", errors.New("expected https://www.google.com/goto?url=...")
	}

	// Default Client follows 3xx; stop on the first hop and read Location.
	client := &http.Client{
		CheckRedirect: func(*http.Request, []*http.Request) error {
			return http.ErrUseLastResponse
		},
	}

	for _, method := range []string{http.MethodHead, http.MethodGet} {
		req, err := http.NewRequest(method, raw, nil)
		if err != nil {
			return "", err
		}
		// Referer matches a real search click; some hops expect it.
		req.Header.Set("Referer", "https://www.google.com/search")

		res, err := client.Do(req)
		if err != nil {
			return "", err
		}
		location := res.Header.Get("Location")
		_ = res.Body.Close()
		if location == "" {
			continue
		}
		// Location may be relative; resolve it against the request URL.
		abs, err := res.Request.URL.Parse(location)
		if err != nil {
			return "", err
		}
		return abs.String(), nil
	}
	return "", errors.New("no Location header")
}

func main() {
	if len(os.Args) != 2 {
		fmt.Fprintln(os.Stderr, "Usage: go run . 'https://www.google.com/goto?url=CAES...'")
		os.Exit(2)
	}
	dest, err := resolveGoto(os.Args[1])
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}
	fmt.Println(dest)
}

File: examples/javascript/resolve_goto.js
#!/usr/bin/env node
/**
 * Resolve a Google /goto URL by reading the Location header.
 *
 * Google Search results often wrap destinations as:
 *   https://www.google.com/goto?url=CAES...
 * The CAES blob is opaque — you cannot decode it. Google only reveals the
 * real URL in the HTTP Location header.
 *
 * Use fetch with redirect: "manual". Do not follow the hop. Try HEAD
 * first, then GET. Google sometimes returns 402 (or another non-3xx)
 * with Location: still read the header. In Node, Location is visible.
 *
 * Usage:
 *   node examples/javascript/resolve_goto.js 'https://www.google.com/goto?url=CAES...'
 */

function isGoogleGoto(raw) {
  try {
    const url = new URL(raw);
    return (
      url.protocol === "https:" &&
      (url.hostname === "google.com" || url.hostname === "www.google.com") &&
      (url.pathname === "/goto" || url.pathname.startsWith("/goto/"))
    );
  } catch {
    return false;
  }
}

async function resolveGoto(url) {
  if (!isGoogleGoto(url)) {
    throw new Error("expected https://www.google.com/goto?url=...");
  }

  // Referer matches a real search click; some hops expect it.
  const headers = { Referer: "https://www.google.com/search" };

  for (const method of ["HEAD", "GET"]) {
    const res = await fetch(url, {
      method,
      redirect: "manual",
      headers,
    });
    const location = res.headers.get("location");
    if (location) {
      // Location may be relative; resolve it against the request URL.
      return new URL(location, url).href;
    }
  }
  return null;
}

const input = process.argv[2];
if (!input) {
  console.error(
    "Usage: node resolve_goto.js 'https://www.google.com/goto?url=CAES...'",
  );
  process.exit(2);
}

resolveGoto(input)
  .then((dest) => {
    if (!dest) {
      console.error("no Location header");
      process.exit(1);
    }
    console.log(dest);
  })
  .catch((err) => {
    console.error(err.message);
    process.exit(1);
  });
